Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py

```diff
@@ -88,72 +148,206 @@ def PyExec(self):
-            Integration(InputWorkspace='__data_tmp', OutputWorkspace='__data_tmp', EnableLogging=False)
-        ConvertSpectrumAxis(InputWorkspace='__data_tmp', Target=target, EFixed=eFixed, OutputWorkspace=outWS, EnableLogging=False)
-        Transpose(InputWorkspace=outWS, OutputWorkspace=outWS, EnableLogging=False)
-        ResampleX(InputWorkspace=outWS, OutputWorkspace=outWS, XMin=xMin, XMax=xMax, NumberBins=numberBins, EnableLogging=False)
-
-        if cal is not None:
-            ExtractUnmaskedSpectra(InputWorkspace=cal, MaskWorkspace='__mask_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            if isinstance(mtd['__cal_tmp'], IEventWorkspace):
-                Integration(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            CopyInstrumentParameters(data, '__cal_tmp', EnableLogging=False)
-            ConvertSpectrumAxis(InputWorkspace='__cal_tmp', Target=target, EFixed=eFixed, OutputWorkspace='__cal_tmp', EnableLogging=False)
-            Transpose(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            ResampleX(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', XMin=xMin, XMax=xMax, NumberBins=numberBins,
-                      EnableLogging=False)
-            Divide(LHSWorkspace=outWS, RHSWorkspace='__cal_tmp', OutputWorkspace=outWS, EnableLogging=False)
-            if normaliseBy == "Monitor":
-                cal_scale = cal.run().getProtonCharge()
-            elif normaliseBy == "Time":
-                cal_scale = cal.run().getLogData('duration').value
-
-        Scale(InputWorkspace=outWS, OutputWorkspace=outWS, Factor=cal_scale/data_scale, EnableLogging=False)
-
-        if bkg is not None:
-            ExtractUnmaskedSpectra(InputWorkspace=bkg, MaskWorkspace='__mask_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            if isinstance(mtd['__bkg_tmp'], IEventWorkspace):
-                Integration(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            CopyInstrumentParameters(data, '__bkg_tmp', EnableLogging=False)
-            ConvertSpectrumAxis(InputWorkspace='__bkg_tmp', Target=target, EFixed=eFixed, OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            Transpose(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            ResampleX(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', XMin=xMin, XMax=xMax, NumberBins=numberBins,
-                      EnableLogging=False)
+        _get_scale = lambda x: {
+            None: 1,
+            "Monitor": x.run().getProtonCharge(),
+            "Time": x.run().getLogData("duration").value,
+        }[normaliseBy]
+
+        # NOTE:
+        # StringArrayProperty cannot be optional, so the background can only be passed in as a string
+        # or a list, which will be manually unpacked here
+        try:
+            bkg = bkg.split(",") if bkg is not None else bkg
+        except:
```

Comment:
It is better to catch specific error.

Reply:
Since the try-except block here is more of a workaround for not being able to use `StringArrayProerty`, I think it is better to let the whole processing fail hard early (upon loading/reading data) than sometime later.
Once `StringArrayProperty` can be an optional input, this whole block should simply go away.